Loader/SymfonyFixturesLoader.php

```diff
@@ -72,6 +79,46 @@ protected function createFixture($class)
         return $this->loadedFixtures[$class];
     }
 
+    /**
+     * Returns the array of data fixtures to execute.
+     *
+     * @return array $fixtures
+     */
+    public function getFixtures($set = null)
+    {
+        $fixtures = parent::getFixtures();
+
+        if ($set) {
+            $mapping = $this->setsFixtureMapping;
+            $fixtures = array_filter(
+                $fixtures,
+                function ($fixture) use ($mapping, $set) {
+                    return isset($mapping[$set]) && in_array(get_class($fixture), $mapping[$set]);
+                }
+            );
+        }
+
+        return $fixtures;
+    }
+
+    /**
+     * Creates an array of the sets and there fixtures
+     *
+     * @param array $services
+     *
+     * @return void
+     */
+    public function addSetsFixtureMapping(array $services)
+    {
+        foreach ($services as $service => $tags) {
+            foreach ($tags as $attributes) {
+                if (array_key_exists('set', $attributes)) {
+                    $this->setsFixtureMapping[$attributes['set']][] = $service;
```

Comment:
This can be:
```
   $this->setsFixtureMapping[$attributes['set']][$service] = true;
```

That change will make sure each service is only stored once, and later you can use `isset` instead of `in_array` (in_array is slow)

Reply:
As another thought, I do not think you should use the service name here.
When you look it up in the method `getFixtures` you use the class name of the service. This will normally be the same in symfony 4 projects, but it does not have to be the same. The service name could easily be something else than the class name of the object implementing the service.

To be on the safe side, I guess you need to either make sure you store the class name as a key when you store the tags, or lookup by service name instead of class name when you do the lookup.

Reply:
Oke i will look into using the get_class name.
Do you happen to know where in symfony the Reference class gets replaced with its other class to use ? [here](https://github.com/doctrine/DoctrineFixturesBundle/blob/master/DependencyInjection/CompilerPass/FixturesCompilerPass.php#L24)
first its an array of References but in the addFixtures method its an array of Fixtures.
If they get created from the id in new Reference($id) will the issue you describe ever happen ?

Reply:
Well, during the compiler pass you have just a "Reference" to the service object. Those references are resolved when actually creating those services after all the compiler passes. So in the  service constructors and injection methods (like our "addFixtures" injection method) those references in the container being injected has been resolved to actual objects representing the services.

In symfony 4 they have the convention that the id of a service should be the fully qualified name of the class implementing the service. But it is only a convention, and before symfony 4 services usually had id's that looked something like "namespace.foo.bar.service". So the issue is very likely to happen.

Reply:
The problem is that the tags get passed to the method "addFixtures" with only a service id linking the tag information to the service that has been tagged. And the services themselves are passed in a different array (in the first argument `$fixtures`) with the resolved services as values. And the method "addFixtures" has no way to find a connection between a service id and those resolved service objects.

I think the solution is that the connection between service reference and tag info has to be maintained in the compiler pass before "addFixtures" is called.
May be pass as an argument just one array, where each element in the array is a 2 element array with the service and the tags info.